### ISSUE TITLE
Keep stack trace: global optimizers in tensor/opt.py

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#4179]
Sub-ticket of #3018.

**Comment by vmichals** [comment]
Are there only these three in tensor/opt.py? [This](http://deeplearning.net/software/theano/extending/optimization.html) says that global optimizers are objects which define apply/add_requirements/optimize methods. I only found these three defining an apply method:
- [ ] ShapeOptimizer
- [ ] UnShapeOptimizer
- [ ] FusionOptimizer

**Comment by lamblin** [comment]
Indeed, these seem to be the only ones. I don't think the first two directly deal with node replacements, so there may be nothing to do.

**Comment by lamblin** [comment]
So, actually, for the FusionOptimizer, the logic should be in the local optimizer it encapsulate.
For `local_add_mul_fusion`, the logic appears to already be there.
For `composite_elemwise_fusion`, we should gather the stack traces of all the intermediate values, in addition to the output, and copy them all into the replacement.
